cmd/gh/main.go

```diff
@@ -106,117 +98,19 @@ func mainRun() exitCode {
-			exe, err := safeexec.LookPath(expandedArgs[0])
-			if err != nil {
-				fmt.Fprintf(stderr, "failed to run external command: %s", err)
-				return exitError
-			}
-
-			externalCmd := exec.Command(exe, expandedArgs[1:]...)
-			externalCmd.Stderr = os.Stderr
-			externalCmd.Stdout = os.Stdout
-			externalCmd.Stdin = os.Stdin
-			preparedCmd := run.PrepareCmd(externalCmd)
-
-			err = preparedCmd.Run()
-			if err != nil {
-				var execError *exec.ExitError
-				if errors.As(err, &execError) {
-					return exitCode(execError.ExitCode())
-				}
-				fmt.Fprintf(stderr, "failed to run external command: %s\n", err)
-				return exitError
-			}
-
-			return exitOK
-		} else if len(expandedArgs) > 0 && !hasCommand(rootCmd, expandedArgs) {
-			extensionManager := cmdFactory.ExtensionManager
-			if found, err := extensionManager.Dispatch(expandedArgs, os.Stdin, os.Stdout, os.Stderr); err != nil {
-				var execError *exec.ExitError
-				if errors.As(err, &execError) {
-					return exitCode(execError.ExitCode())
-				}
-				fmt.Fprintf(stderr, "failed to run extension: %s\n", err)
-				return exitError
-			} else if found {
-				return exitOK
-			}
-		}
-	}
-
-	// provide completions for aliases and extensions
-	rootCmd.ValidArgsFunction = func(cmd *cobra.Command, args []string, toComplete string) ([]string, cobra.ShellCompDirective) {
-		var results []string
-		aliases := cfg.Aliases()
-		for aliasName, aliasValue := range aliases.All() {
-			if strings.HasPrefix(aliasName, toComplete) {
-				var s string
-				if strings.HasPrefix(aliasValue, "!") {
-					s = fmt.Sprintf("%s\tShell alias", aliasName)
-				} else {
-					aliasValue = text.Truncate(80, aliasValue)
-					s = fmt.Sprintf("%s\tAlias for %s", aliasName, aliasValue)
-				}
-				results = append(results, s)
-			}
-		}
-		for _, ext := range cmdFactory.ExtensionManager.List() {
-			if strings.HasPrefix(ext.Name(), toComplete) {
-				var s string
-				if ext.IsLocal() {
-					s = fmt.Sprintf("%s\tLocal extension gh-%s", ext.Name(), ext.Name())
-				} else {
-					path := ext.URL()
-					if u, err := git.ParseURL(ext.URL()); err == nil {
-						if r, err := ghrepo.FromURL(u); err == nil {
-							path = ghrepo.FullName(r)
-						}
-					}
-					s = fmt.Sprintf("%s\tExtension %s", ext.Name(), path)
-				}
-				results = append(results, s)
-			}
-		}
-		return results, cobra.ShellCompDirectiveNoFileComp
-	}
-
-	authError := errors.New("authError")
-	rootCmd.PersistentPreRunE = func(cmd *cobra.Command, args []string) error {
-		// require that the user is authenticated before running most commands
-		if cmdutil.IsAuthCheckEnabled(cmd) && !cmdutil.CheckAuth(cfg) {
-			fmt.Fprint(stderr, authHelp())
-			return authError
-		}
-
-		return nil
+	// translate `gh help <command>` to `gh <command> --help` for extensions.
+	if len(expandedArgs) >= 2 && expandedArgs[0] == "help" && isExtensionCommand(rootCmd, expandedArgs[1:]) {
+		expandedArgs = expandedArgs[1:]
+		expandedArgs = append(expandedArgs, "--help")
```

Comment:
This looks like it should be able to be simplified but my brain has not found a way to yet.